### PR TITLE
AVRO-3633: [Rust] Additional attributes for 'avro_derive' crate

### PR DESCRIPTION
**Pull request by jpramosi**
## What is the purpose of the change

[AVRO-3633](https://issues.apache.org/jira/browse/AVRO-3633) *This pull request introduces additional macro attributes 'skip' and 'rename' to the 'avro_derive' crate.*

## Verifying this change

This change added tests and can be verified as follows:

- *Added test that validates correct behavior for skipping or renaming fields with derive macro 'AvroSchema'.*

## Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? not documented

### CODE REVIEW

File: lang/rust/avro_derive/src/lib.rs [hunk 1]
```diff
@@ -117,11 +121,20 @@ fn get_data_struct_schema_def(
...
-            for (position, field) in a.named.iter().enumerate() {
-                let name = field.ident.as_ref().unwrap().to_string(); // we know everything has a name
+            let mut index: usize = 0;
+            for field in a.named.iter() {
+                let mut name = field.ident.as_ref().unwrap().to_string(); // we know everything has a name
                 let field_attrs =
                     FieldOptions::from_attributes(&field.attrs[..]).map_err(darling_to_syn)?;
                 let doc = preserve_optional(field_attrs.doc);
+                let rename = field_attrs.rename;
+                if rename.is_some() {
```
Comment (jklamer) [inline lang/rust/avro_derive/src/lib.rs:131]: Probably best to use `if let` syntax here as in:
```
if let Some(rename) = rename {
   name = rename.clone()
}
```

File: lang/rust/avro_derive/src/lib.rs [hunk 2]
```diff
@@ -117,11 +121,20 @@ fn get_data_struct_schema_def(
...
+                    name = rename.unwrap();
+                }
+                let skip = field_attrs.skip;
+                if skip.is_some() && skip.unwrap() {
```
Comment (jklamer) [inline lang/rust/avro_derive/src/lib.rs:135]: Same here
```
if let Some(true) = skip {
continue;
}
```
^^ pretty sure that works.

File: lang/rust/avro_derive/tests/derive.rs [hunk 3]
```diff
@@ -1362,4 +1362,144 @@ mod test_derive {
             myenum: MyEnum::Bar,
         });
     }
+
+    #[test]
+    fn test_basic_struct_with_skip_attribute() {
```
Comment (jklamer) [inline lang/rust/avro_derive/tests/derive.rs:1367]: nice testing

**Review by jklamer — COMMENTED** [review]
Great stuff. LGTM. Just a couple nits. I remember not including these because I had some hope that I would be able to follow up with serde and get them to publish their attribute handlers with the library, such that it could all be keyed off JUST the `#[serde(skip)]`. But I don't think thats the obviously right answer in the long run, and if its useful for you now then I'm very happy for this change. @martin-g can give final approval

File: lang/rust/avro_derive/tests/derive.rs (hunk 3 above)
Comment (martin-g) [inline lang/rust/avro_derive/tests/derive.rs:1367]:
```suggestion
    fn avro_3633_test_basic_struct_with_skip_attribute() {
```

File: lang/rust/avro_derive/tests/derive.rs [hunk 4]
```diff
@@ -1362,4 +1362,144 @@ mod test_derive {
...
+                    "map" => panic!("Unexpected field 'map'"),
+                    _ => {}
+                }
+            }
+        } else {
+            panic!("TestBasicStructWithSkipAttribute schema must be a record schema")
+        }
+        assert_eq!(schema, TestBasicStructWithSkipAttribute::get_schema());
+
+        // Note: If serde its skip attribute is used on a field, the field its type
```
Comment (martin-g) [inline lang/rust/avro_derive/tests/derive.rs:1434]:
```suggestion
        // Note: If serde's `skip` attribute is used on a field, the field's type
```

File: lang/rust/avro_derive/tests/derive.rs [hunk 5]
```diff
@@ -1362,4 +1362,144 @@ mod test_derive {
...
+            a: 987.654,
+            map: HashMap::default(), // <- skipped
+            array: vec![4, 5, 6],
+            mystruct: TestBasicStructNoSchema::default(), // <- skipped
+            b: 321,
+        });
+    }
+
+    #[test]
+    fn test_basic_struct_with_rename_attribute() {
```
Comment (martin-g) [inline lang/rust/avro_derive/tests/derive.rs:1449]:
```suggestion
    fn avro_3633_test_basic_struct_with_rename_attribute() {
```

File: lang/rust/avro_derive/tests/derive.rs [hunk 6]
```diff
@@ -1362,4 +1362,144 @@ mod test_derive {
...
+            for field in fields {
+                match field.name.as_str() {
+                    "condition" => panic!("Unexpected field 'condition'"),
+                    "mystruct" => panic!("Unexpected field 'mystruct'"),
+                    "map" => panic!("Unexpected field 'map'"),
+                    _ => {}
+                }
+            }
+        } else {
+            panic!("TestBasicStructWithSkipAttribute schema must be a record schema")
```
Comment (martin-g) [inline lang/rust/avro_derive/tests/derive.rs:1430]: I like the error messages to print the problematic data:

```suggestion
            panic!("TestBasicStructWithSkipAttribute schema must be a record schema: {:?}", derived_schema)
```
You will need to define `derived_schema` as a local variable before the `if let`

File: lang/rust/avro_derive/tests/derive.rs [hunk 7]
```diff
@@ -1362,4 +1362,144 @@ mod test_derive {
...
+            assert_eq!("TestBasicStructWithRenameAttribute", name.fullname(None));
+            for field in fields {
+                match field.name.as_str() {
+                    "a" => panic!("Unexpected field name 'a': must be 'a1'"),
+                    "c" => panic!("Unexpected field name 'c': must be 'c1'"),
+                    _ => {}
+                }
+            }
+        } else {
+            panic!("TestBasicStructWithRenameAttribute schema must be a record schema")
```
Comment (martin-g) [inline lang/rust/avro_derive/tests/derive.rs:1495]:
```suggestion
            panic!("TestBasicStructWithRenameAttribute schema must be a record schema: {:?}", derived_schema)
```

**Review by martin-g — APPROVED** [review]

**Comment by martin-g** [comment]
Thank you for the contribution, @jpramosi !
Thank you for the review, @jklamer !

**Pull request merged**